Common > Validation: 유효성 응답 모델에 "string" 타입 프로퍼티 모델 추가

merge-simpson opened the pull request:
<!--
<br /><br /><br /><br /><br />

<table align="center" border="30">
<tr>
  <td>🏗️ 선행 작업 병합 후, 리베이스할 예정입니다.</td>
</tr>
</table>

<br /><br /><br /><br /><br />
그니까 아직 리뷰하지 말라니깐용.
<br /><br /><br /><br /><br />
<br /><br /><br /><br /><br />
-->

> 이 PR을 리뷰할 경우 #326 PR을 리뷰하지 않아도 됩니다.

## Issues

- Resolves #327 

<br />

## Description

- 문자열 데이터의 유효성 모델을 추가합니다.
- Rel: #323 

<br />

## Review Points

- 노션 스펙을 참고하실 수 있습니다. (로그인 필요)  
  https://www.notion.so/nettee/Auth-User-API-25001973b23a800c9e9fe2392aa7fa97#25001973b23a802aad45ea139d9f1743
- 관련 파일을 참고하실 수 있습니다.  
  - common/src/main/java/nettee/common/validation/model/interfaces/**BaseValidationProperty**.java  
    ```java
    public interface BaseValidationProperty {
        String type();
        Boolean required();
        Map<String, String> messages();
        
        // ... (중략)
    }
    ```  
    ```java
    public interface LengthValidationProperty extends BaseValidationProperty {
        Integer minLength();
        Integer maxLength();
    }
    ```
  - common/src/main/java/nettee/common/validation/model/**ValidationResponseModel**.java

<br />

## How Has This Been Tested?

- 테스트 생략

<br />

## Additional Notes

### 다이어그램

```mermaid
classDiagram
    class BaseValidationProperty {
        &lt;&lt;nterface&gt;&gt;
        +String type()
        +Boolean required()
        +Map~String,String~ messages()
    }

    class LengthValidationProperty {
        &lt;&lt;nterface&gt;&gt;
        +Integer minLength()
        +Integer maxLength()
    }
    
    class StringValidationProperty {
        &lt;&lt;record&gt;&gt;
        +String type()
        +Boolean required()
        +Map~String,String~ messages()
        +Integer minLength()
        +Integer maxLength()
    }

    BaseValidationProperty <|-- LengthValidationProperty
    LengthValidationProperty <|.. StringValidationProperty
```

merge-simpson committed:
feat(common): add base interface for validation response model property

Introduce `BaseValidationProperty` interface to define common contract for validation response.

merge-simpson committed:
feat(common): add reserved field names and expected types for validation model

Define constants in `ReservedFieldNames` and `ExpectedTypes` to standardize validation property schema.

merge-simpson committed:
feat(common): add validation response model

Introduce `ValidationResponseModel` record to encapsulate validation properties with server time and last updated timestamp.
Automatically fallback to current time when server time is `null`.

merge-simpson committed:
feat(common): append reserved field names about datetime fields for validation model

Define constants in `ReservedFieldNames` to standardize validation property schema.

- `MIN_DURATION`
- `MAX_DURATION`
- `STEP`
- `STEP_EPOCH`

merge-simpson committed:
feat(common): append reserved field name "regexp" for validation model

Define constants in `ReservedFieldNames` to standardize validation property schema.

- `REGEXP`

merge-simpson committed:
feat(common): add `LengthValidationProperty` interface for string type validation model

Introduce `LengthValidationProperty` extending `BaseValidationProperty`
to support `minLength` and `maxLength` attributes for string validation.

merge-simpson committed:
feat(common): add string validation property model

implement `stringvalidationproperty` for validation response model; implement `lengthvalidationproperty`; enforce `type` be 'string' with assertion and defaulting; normalize blank `regexp` to `null`; validate non-negative length bounds and ensure `minlength <= maxlength`; auto-initialize messages map and populate defaults for `required`/`regexp`/`length` keys using reserved field names; finalize `messages` as unmodifiable map and log helpful warnings.

wowddok99 commented on common/src/main/java/nettee/common/validation/model/StringValidationProperty.java:
```diff
@@ -0,0 +1,72 @@
+            messages.put(REQUIRED, "필수 입력 항목입니다.");
+        }
+
+        if (regexp != null && !regexp.isBlank() && messages.containsKey(REGEXP)) {
```
정규식(regexp)은 있는데, 개발자가 별도의 안내 메시지를 설정하지 않았을 경우 기본 메시지를 넣어주려는 의도로 보입니다. 
현재는 REGEXP가 포함되어 있을 때만 동작하도록 되어 있는데, 오히려 메시지가 없을 때 기본 메시지를 넣어주는 게 맞을 것 같습니다.

```java
// 정규식은 있지만, 메시지를 별도로 지정하지 않았을 경우 기본 메시지를 부여
if (regexp != null && !regexp.isBlank() && !messages.containsKey(REGEXP)) {
    log.warn("정규표현식에 대한 기본 안내 메시지를 설정합니다."); // 로그 메시지도 함께 명확하게
    messages.put(REGEXP, "올바른 입력 양식이 아닙니다.");
}
```
- !messages.containsKey(REGEXP)로 조건을 변경하면 의도하신 대로 동작할 것 같습니다!!!

merge-simpson replied:
확인 감사합니다. 누락된 느낌표를 모두 추가하였습니다!

merge-simpson committed:
fix(common): correct initializing condition in string validation property model

```diff
-if (~ && messages.containsKey(MAX_LENGTH)) {
+if (~ && !messages.containsKey(MAX_LENGTH)) {
    message.put(key, "default message");
}
```
